docs/needs-restarting.1
.\" needs-restarting
.TH "needs-restarting" "1" "13 January 2013" "" ""
.SH "NAME"
needs-restarting \- report running processes that have been updated
.SH "SYNOPSIS"
\fBneeds-restarting\fP [options]
.SH "DESCRIPTION"
.PP
\fBneeds-restarting\fP is a program that reports a list of process ids that
started running before they or some component that they use were updated.
.PP
.SH "OPTIONS"
.IP "\fB\-h, \-\-help\fP"
Display a help message, and then quit.
.IP "\fB\-u, \-\-useronly\fP"
Show processes for my userid only.
.IP "\fB\-r, \-\-reboothint\fP"
Only report whether a full reboot is required (exit code 1) or not (exit code 0).
.IP "\fB\-s, \-\-services\fP"
List the affected systemd services only.

.PP
.SH "SEE ALSO"
.nf
.I yum.conf (5)
http://yum.baseurl.org/
.fi

.PP
.SH "AUTHORS"
.nf
See the Authors file included with this program.
.fi


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

needs-restarting.py
#!/usr/bin/python -tt

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
# Copyright 2009 Red Hat Inc
# written by Seth Vidal

# look through list of running apps
# report any app which was updated after it was started
# (and therefore needs to be restarted)


# for each /proc/number-dir 
# get stat of create time on that dir
# open up smaps and search for all lines with 'fd:' in them
# take filename
# search for the package owning that file
# make a list of installtimes of all pkgs of the files the program has open
# sort the list
# if the dir create time is < the largest time on the installtimes list
# then output the pid and process cmdline as needing to be restarted

#TODO:
# maybe note deleted files which are not owned by any pkg but which an app
# is still using
# output userids, too?


import sys
import os
import yum
import yum.misc
import glob
import stat
from optparse import OptionParser
from yum.Errors import RepoError
sys.path.insert(0,'/usr/share/yum-cli')
import utils

# For which package updates we should recommend a reboot
# Taken from https://access.redhat.com/solutions/27943
REBOOTPKGS = ['kernel', 'glibc', 'linux-firmware', 'systemd', 'udev',
              'openssl-libs', 'gnutls', 'dbus']

def parseargs(args):
    usage = """
    needs-restarting: Report a list of process ids of programs that started 
                    running before they or some component they use were updated.
    """
    parser = OptionParser(usage=usage)
    
    parser.add_option("-u", "--useronly", default=False, action="store_true",
      help='show processes for my userid only')
    parser.add_option("-r", "--reboothint", default=False, action="store_true",
      help=('only report whether a full reboot is required (exit code 1) or '
            'not (exit code 0)'))
    parser.add_option("-s", "--services", default=False, action="store_true",
      help='list the affected systemd services only')
    
    (opts, args) = parser.parse_args(args)
    return (opts, args)


def return_running_pids(uid=None):
    mypid = os.getpid()
    pids = []
    for fn in glob.glob('/proc/[0123456789]*'):
        if mypid == os.path.basename(fn):
            continue

        if uid: # meaning we're not root and we've added -u
            if os.stat(fn)[stat.ST_UID] != uid:
                continue

        pids.append(os.path.basename(fn))
    return pids

def get_open_files(pid):
    files = []
    smaps = '/proc/%s/smaps' % pid
    try:
        with open(smaps, 'r') as maps_f:
            maps = maps_f.readlines()
    except (IOError, OSError), e:
        print >>sys.stderr, "Could not open %s" % smaps
        return files

    for line in maps:
        slash = line.find('/')
        if slash == -1 or line.find('00:') != -1: # if we don't have a '/' or if we fine 00: in the file then it's not _REALLY_ a file
            continue
        line = line.replace('\n', '')
        filename = line[slash:]
        filename = filename.split(';')[0]
        filename = filename.strip()
        if filename not in files:
            files.append(filename)
    return files

def get_service(pid):
    """Return the systemd service to which the process belongs.

    More details:
    http://0pointer.de/blog/projects/systemd-for-admins-2.html
    https://www.freedesktop.org/wiki/Software/systemd/FrequentlyAskedQuestions/
    """

    fname = '/proc/%s/cgroup' % pid
    try:
        with open(fname, 'r') as f:
            groups = f.readlines()
    except (IOError, OSError), e:
        print >>sys.stderr, "Could not open %s" % fname
        return None

    for line in groups:
        line = line.replace('\n', '')
        hid, hsub, cgroup = line.split(':')
        if hsub == 'name=systemd':
            name = cgroup.split('/')[-1]
            if name.endswith('.service'):
                return name
    return None

def main(args):
    (opts, args)  = parseargs(args)

    myuid = os.geteuid()
    if myuid and not opts.useronly:
        print >>sys.stderr, "Please run as root or with --useronly option"
        sys.exit(1)

    my = yum.YumBase()
    my.preconf.init_plugins=False
    if hasattr(my, 'setCacheDir'):
        my.setCacheDir()
    my.conf.cache = True
    
    boot_time = utils.get_boot_time()

    if opts.reboothint:
        needing_reboot = set()
        for pkg in my.rpmdb.searchNames(REBOOTPKGS):
            if float(pkg.installtime) > float(boot_time):
                needing_reboot.add(pkg)
        if needing_reboot:
            print 'Core libraries or services have been updated:'
            for pkg in needing_reboot:
                print '  %s ->' % pkg.name, pkg.printVer()
            print
            print 'Reboot is required to ensure that your system benefits',
            print 'from these updates.'
            print
            print 'More information:'
            print 'https://access.redhat.com/solutions/27943'
            return 1
        else:
            print 'No core libraries or services have been updated.'
            print 'Reboot is probably not necessary.'
            return 0

    needing_restart = set()
    for pid in return_running_pids(uid=myuid):
        try:
            pid_start = utils.get_process_time(int(pid), boot_time)['start_time']
        except (OSError, IOError), e:
            continue
        found_match = False
        for fn in get_open_files(pid):
            if found_match:
                break
            just_fn = fn.replace('(deleted)', '')
            just_fn = just_fn.strip()            
            bogon = False
            # if the file is in a pkg which has been updated since we started the pid - then it needs to be restarted            
            for pkg in my.rpmdb.searchFiles(just_fn):
                if float(pkg.installtime) > float(pid_start):
                    needing_restart.add(pid)
                    found_match = True
                    continue
                if just_fn in pkg.ghostlist:
                    bogon = True
                    break
            
            if bogon:
                continue

            # if the file is deleted 
            if fn.find('(deleted)') != -1: 
                # and it is from /*bin/* then it needs to be restarted 
                if yum.misc.re_primary_filename(just_fn):
                    needing_restart.add(pid)
                    found_match = True
                    continue

                # if the file is from an old ver of an installed pkg - then assume it was just updated but the 
                # new pkg doesn't have the same file names. Fabulous huh?!
                my.conf.cache = False
                for oldpkg in my.pkgSack.searchFiles(just_fn): # ghostfiles are always bogons
                    if just_fn in oldpkg.ghostlist:
                        continue
                    if my.rpmdb.installed(oldpkg.name):
                        needing_restart.add(pid)
                        found_match = True
                        break

           
            
    if opts.services:
        names = set([get_service(pid) for pid in needing_restart])
        for name in names:
            if name is not None:
                print name
        return 0

    for pid in needing_restart:
        try:
            cmdline = open('/proc/' +pid+ '/cmdline', 'r').read()
        except (OSError, IOError), e:
            print >>sys.stderr, "Couldn't access process information for %s: %s" % (pid, str(e))
            continue
        # proc cmdline is null-delimited so clean that up
        cmdline = cmdline.replace('\000', ' ')
        print '%s : %s' % (pid, cmdline)
        
    return 0
    
if __name__ == "__main__":
    try:
        sys.exit(main(sys.argv))
    except RepoError, e:
        print >>sys.stderr, e
        sys.exit(1)
